src/sync.cpp

```diff
@@ -340,18 +337,16 @@ void CSharedCriticalSection::unlock_shared()
             printf("never locked at %s:%d\n", li.file, li.line);
             assert(alreadyLocked != sharedowners.end());
         }
-    }
-    boost::shared_mutex::unlock_shared();
-    {
-        boost::unique_lock<boost::mutex> lock(setlock);
         sharedowners.erase(tid);
     }
+    boost::shared_mutex::unlock_shared();
 }
 
 bool CSharedCriticalSection::try_lock_shared()
 {
     // detect recursive locking
     uint64_t tid = getTid();
+    boost::unique_lock<boost::mutex> lock(setlock);
     assert(exclusiveOwner != tid);
```

Comment:
why not use std::unique_lock , std::mutex?

Reply:
I will.  This is merged code so was written before we moved to c++11